File: 4033-valid-k-unique-subarrays-i/4033-valid-k-unique-subarrays-i.go
package main

import (
	"math"
	"sort"
)

type counter struct {
	mp   map[int]int
	odds int
}

func (c *counter) init() {
	if c.mp == nil {
		c.mp = make(map[int]int)
	}
}

func (c *counter) update(x int, op int) {
	c.init()
	c.mp[x] += op
	if c.mp[x]&1 == 1 {
		c.odds++
	} else {
		c.odds--
	}
	if c.mp[x] == 0 {
		delete(c.mp, x)
	}
}

func (c *counter) ok(k int) bool {
	if len(c.mp) != k {
		return false
	}
	return c.odds == 0
}

type q3 struct {
	l   int
	r   int
	idx int
}

func validSubarrays(nums []int, k int, queries [][]int) []bool {
	var c counter
	Q := make([]q3, 0)

	for i := 0; i < len(queries); i++ {
		l := queries[i][0]
		r := queries[i][1]
		length := r - l + 1
		if length&1 == 1 {
			continue
		}
		if length < 2*k {
			continue
		}
		Q = append(Q, q3{l: l, r: r, idx: i})
	}

	sq := int(math.Sqrt(float64(len(queries))))
	if sq == 0 {
		sq = 1
	}

	sort.Slice(Q, func(i, j int) bool {
		a := Q[i]
		b := Q[j]
		ab := a.l / sq
		bb := b.l / sq
		if ab == bb {
			return a.r < b.r
		}
		return ab < bb
	})

	res := make([]bool, len(queries))
	l, r := 0, 0

	for _, qu := range Q {
		le, ri, idx := qu.l, qu.r, qu.idx
		for r <= ri {
			c.update(nums[r], 1)
			r++
		}
		for l > le {
			l--
			c.update(nums[l], 1)
		}
		for r > ri+1 {
			r--
			c.update(nums[r], -1)
		}
		for l < le {
			c.update(nums[l], -1)
			l++
		}
		res[idx] = c.ok(k)
	}

	return res
}

File: 4033-valid-k-unique-subarrays-i/4033-valid-k-unique-subarrays-i.cpp

struct Counter {
    unordered_map<int,int> mp;
    int odds = 0;
    void update(int x, int op) {
        mp[x] += op;
        if(mp[x] & 1) odds++;
        else odds--;
        if(mp[x] == 0) mp.erase(x);
    }
    bool ok(int k) {
        if(mp.size() != k) return false;
        return !odds;
    }
};
class Solution {
public:
    vector<bool> validSubarrays(vector<int>& nums, int k, vector<vector<int>>& queries) {
        Counter c;
        vector<array<int,3>> Q;
        vector<int> counts{0};
        unordered_set<int> us;
        for(int i = 0; i < nums.size(); i++) {
            us.insert(nums[i]);
            counts.push_back(us.size());
        }
        for(int i = 0; i < queries.size(); i++) {
            int l = queries[i][0], r = queries[i][1];
            int len = r - l + 1;
            if(len & 1) continue;
            if(len < 2 * k) continue;
            if(counts[r+1] - counts[l] > k) continue;
            if(len % k) continue;
            Q.push_back({l, r, i});
        }
        int sq = sqrt(queries.size());
        sort(begin(Q), end(Q), [&](auto& a, auto& b) {
            int asq = a[0] / sq, bsq = b[0] / sq;
            if (asq == bsq) {
                return a[1] < b[1];
            }
            return asq < bsq;
        });
        vector<bool> res(queries.size());
        int l = 0, r = 0;
        for(auto& [le,ri,idx] : Q) {
            while(r <= ri) c.update(nums[r++],1);
            while(l > le) c.update(nums[--l], 1);
            while(r > ri + 1) c.update(nums[--r],-1);
            while(l < le) c.update(nums[l++], -1);
            res[idx] = c.ok(k);
        }
        return res;
    }
};